cf_il/model.py

```diff
@@ -0,0 +1,89 @@
+from argparse import Namespace
+from typing import Any
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+import torchvision
+
+from cf_il.recover_memory import recovery_memory as rec_mem
+from models.utils.continual_model import ContinualModel
+from utils.buffer import Buffer
+
+
+class CFIL(ContinualModel):
+    NAME: str = 'cf-il'  # type: ignore[assignment]
+    COMPATIBILITY = [
+        'class-il',
+    ]
+
+    def __init__(
+        self,
+        backbone: nn.Module,
+        loss: nn.Module,
+        args: Namespace,
+        transform: torchvision.transforms,
+    ):
+        super(CFIL, self).__init__(backbone, loss, args, transform)
+        # Buffer contains synthetic images from RMP
+        self.buffer = Buffer(
+            buffer_size=self.args.buffer_size,
+            device=self.device,
+            mode='reservoir',
+        )
+
+        self.opt = optim.SGD(
+            backbone.parameters(),
+            lr=self.args.lr,
+            momentum=self.args.momentum,
+        )
+
+    def observe(
+        self,
+        inputs: torch.Tensor,
+        labels: torch.Tensor,
+        *args: Any,
+        **kwargs: Any,
+    ) -> float:
+
+        self.opt.zero_grad()
+
+        outputs = self.net(inputs)
+        loss = self.loss(outputs, labels)
+
+        print(f'Real dataset loss: {loss}')
+
+        if not self.buffer.is_empty():
+            buf_inputs, buf_logits = self.buffer.get_data(self.args.minibatch_size)
+            buf_outputs = self.net(buf_inputs)
+            synth_loss = F.mse_loss(buf_outputs, buf_logits)
+            print(f'Synthetic dataset loss: {synth_loss}')
+            loss += self.args.alpha * synth_loss
+
+        print(f'Total loss: {loss}')
+
+        loss.backward()
+        self.opt.step()
+
+        return loss.item()  # type: ignore[no-any-return]
+
+    def recover_memory(
+        self,
+        num_tasks: int,
+        num_classes_per_task: int = 2,
+        num_images_per_class: int = 10,
+        scale: float = 1,
+    ) -> None:
```

Comment:
Ustawić model w tryb ewaluacji, a następnie wrócić do poprzedniego staniu (tak jak jest to robione w funkcjach `evaluate` w `training.py`), bo tak to chyba będziemy aktualizować wagi podczas destylacji wiedzy.

Reply:
EDIT: chyba jednak nie, bo nie robimy kroku optymatyzatora (ale wciąż przejście w tryb eval chyba może zwiększyć szybkość destylacji)

Reply:
Wydaje się, że chyba jest dobrze. Wagi będą aktualizowane tylko raz na podstawie połączonej straty z klasyfikacji na prawdziwych i destylowanych obrazkach.

Reply:
True, choć `eval` robi jeszcze kilka rzeczy z tego co wyczytałem https://stackoverflow.com/questions/60018578/what-does-model-eval-do-in-pytorch